### PR TITLE
detect/asn1: Migrate to Rust - v9

### PR DESCRIPTION
**Pull request by sharksforarms**
Previous PR: https://github.com/OISF/suricata/pull/5125

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/2977

Describe changes:
- Migration of ASN1 decoder and detect module to Rust

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: rust/Cargo.toml.in [hunk 1]
```diff
@@ -29,12 +29,16 @@ num-derive = "0.2"
...
 ntp-parser = "0.4"
 ipsec-parser = "0.5"
-snmp-parser = "0.5"
+snmp-parser = "0.6"
 tls-parser = "0.9"
 x509-parser = "0.6.5"
 libc = "0.2.67"
+
+[dev-dependencies]
+hex-literal = "0.2"
```
Comment (victorjulien) [inline rust/Cargo.toml.in:43]: where is this used?

File: rust/Cargo.toml.in [hunk 2]
```diff
@@ -29,12 +29,16 @@ num-derive = "0.2"
...
+test-case = "1.0"
```
Comment (jasonish) [inline rust/Cargo.toml.in:44]: Can you tell us why you chose this test helper? We're not opposed to it, but are cautious with pulling in new dependencies, and in particular how aggressive they may be at dropping support for older versions of Rust.

File: rust/src/common.rs [hunk 3]
```diff
@@ -72,3 +72,20 @@ pub unsafe extern "C" fn rs_cstring_free(s: *mut c_char) {
...
+///
+/// # Safety
+///
+/// `ptr` must be a pointer allocated via the Box interface
+pub unsafe extern "C" fn rs_free_boxed<T>(ptr: *mut T) {
+    if ptr.is_null() {
+        return;
+    }
+    drop(Box::from_raw(ptr));
+}
```
Comment (jasonish) [inline rust/src/common.rs:91]: Can you remove these functions and the usage of them for now. I think some generalization may be needed in this area, just not confident on how this should look yet, so I don't want to introduce something like this yet.

File: rust/Cargo.toml.in (hunk 2 above)
Comment (jasonish) [reply rust/Cargo.toml.in:44]: Right now I have 2 concerns with the test-case crate.  Cargo test fails, and their own CI tests in Github also fail.

Comment (sharksforarms) [reply rust/Cargo.toml.in:44]: I chose it in order to introduce parameterized testing, this allows to easily add more tests and to reduce code duplication.

There's no particular reason why this crate over another. I've used both `test-case` and `rstest` in the past. I'm not oppose to changing it either.

File: rust/Cargo.toml.in (hunk 1 above)
Comment (sharksforarms) [reply rust/Cargo.toml.in:43]: Must have used it during local dev. Will be removed in next branch.

File: rust/src/common.rs (hunk 3 above)
Comment (sharksforarms) [reply rust/src/common.rs:91]: No problem. I'll remove the last commit.

**Pull request closed without merging**